File: crates/mathhook-benchmarks/public/node/runAll.js
#!/usr/bin/env node
/**
 * Run All Benchmarks
 * ===================
 *
 * Execute all Node.js benchmark modules and generate a comprehensive report.
 *
 * Usage:
 *     node runAll.js [--json] [--samples N] [--save FILEPATH]
 *
 * Options:
 *     --json          Output results as JSON
 *     --samples N     Number of samples per benchmark (default: 100)
 *     --save PATH     Save results to file
 *
 * Last Updated: 2025-11-29T2300
 */

const path = require('path');
const fs = require('fs');
const { saveResults } = require('./benchmarkUtils');

// Benchmark modules
const corePerformance = require('./corePerformance');
const calculusBenchmarks = require('./calculusBenchmarks');
const solvingBenchmarks = require('./solvingBenchmarks');
const simplificationBenchmarks = require('./simplificationBenchmarks');
const functionEvaluationBenchmarks = require('./functionEvaluationBenchmarks');
const polynomialBenchmarks = require('./polynomialBenchmarks');
const parsingBenchmarks = require('./parsingBenchmarks');

/**
 * Main runner
 */
function main() {
    const args = process.argv.slice(2);
    const jsonOutput = args.includes('--json');
    const samples = args.includes('--samples') ? parseInt(args[args.indexOf('--samples') + 1]) : 100;
    const saveIdx = args.indexOf('--save');
    const savePath = saveIdx !== -1 ? args[saveIdx + 1] : null;

    console.error("=".repeat(80));
    console.error("MathHook Node.js Comprehensive Benchmark Suite");
    console.error("=".repeat(80));
    console.error(`Samples per benchmark: ${samples}`);
    console.error(`Node.js version: ${process.version}`);
    console.error();

    const allResults = {
        platform: 'nodejs-mathhook',
        node_version: process.version,
        timestamp: new Date().toISOString(),
        samples,
        suites: {}
    };

    // Run all benchmark suites
    const suites = [
        { name: 'core_performance', module: corePerformance },
        { name: 'calculus', module: calculusBenchmarks },
        { name: 'solving', module: solvingBenchmarks },
        { name: 'simplification', module: simplificationBenchmarks },
        { name: 'function_evaluation', module: functionEvaluationBenchmarks },
        { name: 'polynomial', module: polynomialBenchmarks },
        { name: 'parsing', module: parsingBenchmarks }
    ];

    suites.forEach(({ name, module }) => {
        console.error(`\nRunning ${name} benchmarks...`);
        console.error("-".repeat(60));

        try {
            // Run the module's main function
            const originalArgv = process.argv;
            process.argv = ['node', 'benchmark', '--json', '--samples', samples.toString()];

            // Capture stdout
            let output = '';
            const originalStdoutWrite = process.stdout.write;
            process.stdout.write = (chunk) => {
                output += chunk;
                return true;
            };

            module.main();

            // Restore stdout
            process.stdout.write = originalStdoutWrite;
            process.argv = originalArgv;

            // Parse results
            allResults.suites[name] = JSON.parse(output);

        } catch (e) {
            console.error(`ERROR in ${name}: ${e.message}`);
            allResults.suites[name] = { error: e.message };
        }
    });

    // Output results
    if (jsonOutput) {
        console.log(JSON.stringify(allResults, null, 2));
    } else {
        printSummary(allResults);
    }

    // Save results if requested
    if (savePath) {
        saveResults(allResults, savePath);
        console.error(`\nResults saved to: ${savePath}`);
    }
}

/**
 * Print human-readable summary of all benchmark results
 */
function printSummary(allResults) {
    console.log("\n" + "=".repeat(80));
    console.log("Benchmark Suite Summary");
    console.log("=".repeat(80));
    console.log(`Node.js version: ${allResults.node_version}`);
    console.log(`Timestamp: ${allResults.timestamp}`);
    console.log(`Samples per benchmark: ${allResults.samples}`);

    Object.entries(allResults.suites).forEach(([suiteName, suiteData]) => {
        console.log(`\n${suiteName.toUpperCase()}:`);
        console.log("-".repeat(60));

        if (suiteData.error) {
            console.log(`  ERROR: ${suiteData.error}`);
            return;
        }

        if (!suiteData.benchmarks) {
            console.log("  No benchmark data");
            return;
        }

        // Calculate statistics
        const benchmarks = Object.values(suiteData.benchmarks).filter(b => !b.error);
        if (benchmarks.length === 0) {
            console.log("  No successful benchmarks");
            return;
        }

        const meanTimes = benchmarks.map(b => b.mean_ns);
        const totalMean = meanTimes.reduce((a, b) => a + b, 0) / meanTimes.length;
        const minTime = Math.min(...meanTimes);
        const maxTime = Math.max(...meanTimes);

        console.log(`  Total benchmarks: ${Object.keys(suiteData.benchmarks).length}`);
        console.log(`  Successful: ${benchmarks.length}`);
        console.log(`  Failed: ${Object.keys(suiteData.benchmarks).length - benchmarks.length}`);
        console.log(`  Average time: ${(totalMean / 1000).toFixed(2)} us`);
        console.log(`  Min time: ${(minTime / 1000).toFixed(2)} us`);
        console.log(`  Max time: ${(maxTime / 1000).toFixed(2)} us`);
    });

    // Overall statistics
    console.log("\n" + "=".repeat(80));
    console.log("Overall Statistics");
    console.log("=".repeat(80));

    let totalBenchmarks = 0;
    let totalSuccessful = 0;
    let totalFailed = 0;
    const allMeanTimes = [];

    Object.values(allResults.suites).forEach(suiteData => {
        if (suiteData.error || !suiteData.benchmarks) return;

        const benchmarkCount = Object.keys(suiteData.benchmarks).length;
        totalBenchmarks += benchmarkCount;

        Object.values(suiteData.benchmarks).forEach(benchmark => {
            if (benchmark.error) {
                totalFailed++;
            } else {
                totalSuccessful++;
                allMeanTimes.push(benchmark.mean_ns);
            }
        });
    });

    console.log(`Total benchmark count: ${totalBenchmarks}`);
    console.log(`Successful: ${totalSuccessful}`);
    console.log(`Failed: ${totalFailed}`);

    if (allMeanTimes.length > 0) {
        const overallMean = allMeanTimes.reduce((a, b) => a + b, 0) / allMeanTimes.length;
        const overallMin = Math.min(...allMeanTimes);
        const overallMax = Math.max(...allMeanTimes);

        console.log(`Overall average time: ${(overallMean / 1000).toFixed(2)} us`);
        console.log(`Overall min time: ${(overallMin / 1000).toFixed(2)} us`);
        console.log(`Overall max time: ${(overallMax / 1000).toFixed(2)} us`);
    }
}

if (require.main === module) {
    main();
}

module.exports = { main };

File: crates/mathhook-benchmarks/public/python/run_all.py
"""
Run All Python Benchmarks

Runs all benchmark categories and outputs combined JSON results.
"""

import argparse
import json
import sys
from typing import Dict, Any

try:
    from mathhook import Expression
except ImportError:
    print("ERROR: mathhook Python bindings not found. Install with: pip install mathhook")
    exit(1)

# Import all benchmark modules
import core_performance
import calculus_benchmarks
import solving_benchmarks
import simplification_benchmarks
import function_evaluation_benchmarks
import polynomial_benchmarks
import parsing_benchmarks


BENCHMARK_CATEGORIES = {
    "core": (core_performance, "Core Performance"),
    "calculus": (calculus_benchmarks, "Calculus"),
    "solving": (solving_benchmarks, "Solving"),
    "simplification": (simplification_benchmarks, "Simplification"),
    "functions": (function_evaluation_benchmarks, "Function Evaluation"),
    "polynomials": (polynomial_benchmarks, "Polynomials"),
    "parsing": (parsing_benchmarks, "Parsing"),
}


def format_result_to_dict(result) -> Dict[str, Any]:
    """Convert BenchmarkResult to dictionary for JSON serialization."""
    return {
        "name": result.name,
        "mean_ns": result.mean_ns,
        "median_ns": result.median_ns,
        "min_ns": result.min_ns,
        "max_ns": result.max_ns,
        "std_dev_ns": result.std_dev_ns,
        "samples": result.samples,
    }


def run_category(category_name: str, module, samples: int = 100) -> Dict[str, Any]:
    """Run benchmarks for a specific category."""
    print(f"\n{'=' * 80}")
    print(f"Running {BENCHMARK_CATEGORIES[category_name][1]} Benchmarks")
    print(f"{'=' * 80}")

    results = module.run_all_benchmarks(samples=samples)

    # Convert to JSON-serializable format
    json_results = {
        name: format_result_to_dict(result)
        for name, result in results.items()
    }

    return json_results


def run_all_benchmarks(categories=None, samples: int = 100) -> Dict[str, Any]:
    """
    Run all or specified benchmark categories.

    Args:
        categories: List of category names to run, or None for all
        samples: Number of samples per benchmark

    Returns:
        Dictionary with all benchmark results
    """
    if categories is None:
        categories = list(BENCHMARK_CATEGORIES.keys())

    all_results = {}

    for category in categories:
        if category not in BENCHMARK_CATEGORIES:
            print(f"WARNING: Unknown category '{category}', skipping...")
            continue

        module, _ = BENCHMARK_CATEGORIES[category]
        category_results = run_category(category, module, samples=samples)
        all_results[category] = category_results

    return all_results


def print_summary(results: Dict[str, Any]):
    """Print summary statistics for all benchmarks."""
    print("\n" + "=" * 80)
    print("BENCHMARK SUMMARY")
    print("=" * 80)

    for category, category_results in results.items():
        category_name = BENCHMARK_CATEGORIES[category][1]
        print(f"\n{category_name}:")
        print("-" * 80)

        # Calculate category statistics
        all_means = [result["mean_ns"] for result in category_results.values()]
        if all_means:
            min_mean = min(all_means)
            max_mean = max(all_means)
            avg_mean = sum(all_means) / len(all_means)

            print(f"  Benchmarks: {len(category_results)}")
            print(f"  Fastest:    {min_mean:.2f}ns")
            print(f"  Slowest:    {max_mean:.2f}ns")
            print(f"  Average:    {avg_mean:.2f}ns")

    print("\n" + "=" * 80)


def main():
    """Main entry point for running all benchmarks."""
    parser = argparse.ArgumentParser(
        description="Run MathHook Python benchmarks",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  %(prog)s                              # Run all benchmarks
  %(prog)s --category core              # Run only core benchmarks
  %(prog)s --category core calculus     # Run core and calculus benchmarks
  %(prog)s --samples 50                 # Run with 50 samples per benchmark
  %(prog)s --output results.json        # Save results to JSON file
        """
    )

    parser.add_argument(
        "--category",
        "-c",
        nargs="+",
        choices=list(BENCHMARK_CATEGORIES.keys()),
        help="Benchmark categories to run (default: all)",
    )

    parser.add_argument(
        "--samples",
        "-s",
        type=int,
        default=100,
        help="Number of samples per benchmark (default: 100)",
    )

    parser.add_argument(
        "--output",
        "-o",
        type=str,
        help="Output JSON file path (optional)",
    )

    parser.add_argument(
        "--quiet",
        "-q",
        action="store_true",
        help="Suppress detailed output, only show summary",
    )

    args = parser.parse_args()

    # Run benchmarks
    results = run_all_benchmarks(categories=args.category, samples=args.samples)

    # Print summary
    if not args.quiet:
        print_summary(results)

    # Save to JSON if requested
    if args.output:
        with open(args.output, "w") as f:
            json.dump(results, f, indent=2)
        print(f"\nResults saved to: {args.output}")

    # Return success
    return 0


if __name__ == "__main__":
    sys.exit(main())
